knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go

```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
 }
 
-type skipDeletePredicate struct {
-	predicate.Funcs
+func (r *ReconcileSourceDeployment) reconcileSourceMonitoring(dep *v1.Deployment, shouldEnableMonitoring bool, inDeletion bool) error {
+	// If monitoring is set to on/off this triggers a global resync to source adapters.
+	// Same applies if we change any of the env vars affecting cluster monitoring or service monitor resource generation.
+	// The Serverless operator pod is restarted and local informer caches are synched.
+	if shouldEnableMonitoring {
+		// If in deletion there is nothing to be done, owner refs will remove source service monitors
+		if !inDeletion {
+			if err := r.setupClusterMonitoringForSources(dep); err != nil {
+				return err
+			}
+			if err := r.generateSourceServiceMonitors(dep); err != nil {
+				return err
+			}
+		}
+	} else {
+		// Remove any relics if previously monitoring was on.
+		if err := r.cleanUpSourceMonitoringResources(dep, inDeletion); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (skipDeletePredicate) Delete(e event.DeleteEvent) bool {
-	return false
+func (r *ReconcileSourceDeployment) generateSourceServiceMonitors(dep *v1.Deployment) error {
+	shouldGenerateSourceMonitors, err := r.shouldGenerateSourceServiceMonitorsByDefault()
+	if err != nil {
+		return err
+	}
+	if shouldGenerateSourceMonitors {
+		if err := SetupSourceServiceMonitorResources(r.client, dep); err != nil {
+			return err
+		}
+	} else {
+		if dep.Namespace != "knative-eventing" {
+			if err := RemoveSourceServiceMonitorResources(r.client, dep); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Setup cluster monitoring Prometheus monitoring requirements
+func (r *ReconcileSourceDeployment) setupClusterMonitoringForSources(dep *v1.Deployment) error {
+	shouldEnableClusterMonitoring, err := r.shouldUseClusterMonitoringForSourcesByDefault()
+	if err != nil {
+		return err
+	}
+	if shouldEnableClusterMonitoring {
+		if err := monitoring.SetupClusterMonitoringRequirements(r.client, dep); err != nil {
+			return err
+		}
+	} else {
+		// Make sure we disable cluster monitoring if we have to eg. we move from a state of enabled to disabled and
+		// resources are left without cleanup. This brings us to the right state.
+		if dep.Namespace != "knative-eventing" {
+			if err := monitoring.RemoveClusterMonitoringRequirements(r.client, dep); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
 }
 
-type skipUpdatePredicate struct {
+// Remove any Source service monitor resources in any user namespace
+func (r *ReconcileSourceDeployment) cleanUpSourceMonitoringResources(dep *v1.Deployment, inDeletion bool) error {
+	if dep.Namespace != "knative-eventing" {
+		if err := monitoring.RemoveClusterMonitoringRequirements(r.client, dep); err != nil {
+			return err
+		}
+		// No need to do anything if in deletion phase as owner references will do the cleanup
+		if !inDeletion {
```

Comment:
Ahhhh, you distinguish between "disable monitoring" and "removing the deployment". Is that distinction necessary? I think we can safely drop the `inDeletion` state and handle both cases the same. It's not harmful to actively delete things even if the ownerRef would technically take care of it.

Reply:
If I get an update event for a source deployment and monitoring is enabled then I want to avoid setting up the source resources (if in the meanwhile someone deleted the source). Note that when I am saying source resources it is not just the service monitors that have owner refs set to the source but also global rbac rules for the source namespace to be used by Prometheus sa that lives in the openshift-monitoring ns. If I setup source monitoring resources and the source is in deletion phase rbac rules will be left undeleted if that is the only source in that namespace. 
That is why I keep the state there because I have the delete predicate that ignores deletes. I can change that to handle deletions but the logic so far was only to handle updates.